hyperspace/parachain/src/finality_protocol.rs

```diff
@@ -362,15 +367,43 @@ where
 		)))?,
 	};
 
-	if justification.commit.target_number <= client_state.latest_relay_height {
-		Err(anyhow!(
-			"skipping outdated commit: {}, with latest relay height: {}",
-			justification.commit.target_number,
-			client_state.latest_relay_height
-		))?
-	}
-
 	let prover = source.grandpa_prover();
+	let (session_start, session_end) =
+		prover.session_start_and_end_for_block(client_state.latest_relay_height).await?;
+	let next_relay_height = if session_end == client_state.latest_relay_height {
+		session_end + 1
+	} else if client_state.latest_relay_height == session_start {
+		session_start + 1
+	} else if client_state.latest_relay_height > session_start &&
+		client_state.latest_relay_height < session_end
+	{
+		client_state.latest_relay_height + 1
+	} else {
+		session_end
+	};
```

Comment:
```suggestion
	let next_relay_height = if client_state.latest_relay_height >= session_start &&
		client_state.latest_relay_height <= session_end
	{
		client_state.latest_relay_height + 1
	} else {
		session_end
	};
```

But probably it can even be simplified to
```rust
let next_relay_height = if client_state.latest_relay_height == session_end {
    session_end + 1
} else {
    session_end
};
```

Reply:
That doesn't work, it omits in between updates and only allows us send session block updates

Reply:
If `client_state.latest_relay_height` is not equal to the session end then we'll miss all other updates because we'll be waiting till the session end, in a case where the session is  4 hours we'll only send an update every four hours.

Reply:
The second suggestion causes us to omit in between updates, it would be noticeable if the session length is large say a couple hours, you didn't notice it in the test because a session is 10 blocks.

Reply:
So `prove_finality` doesn't return all blocks in a session (up to the given one)?

Reply:
Not about `prove_finality`, the relayer would wait until the session_end block is finalized until it sends any update, and we don't want that.

Reply:
I just thought, we call `prover.session_start_and_end_for_block(client_state.latest_relay_height)`. It means that `client_state.latest_relay_height` will always be in between `session_start` and `session_end`, so the condition is always `true`

Reply:
We could just use `client_state.latest_relay_height + 1`